verilog/analysis/checkers/generate_label_prefix_rule.cc

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/generate_label_prefix_rule.h"
+
+#include <vector>
+#include <string>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/match.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/analysis/matcher/matcher.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+#include "verilog/CST/identifier.h"
+#include "verilog/CST/seq_block.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+
+// Register the lint rule
+VERILOG_REGISTER_LINT_RULE(GenerateLabelPrefixRule);
+
+absl::string_view GenerateLabelPrefixRule::Name() { return "generate-label-prefix"; }
+const char GenerateLabelPrefixRule::kTopic[] = "generate-statements";
+const char GenerateLabelPrefixRule::kMessage[] =
+    "All generate block labels must start with g_ or gen_";
+
+std::string GenerateLabelPrefixRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat(
+      "Checks that every generate block label starts with g_ or gen_. See ",
+      GetStyleGuideCitation(kTopic), ".");
+}
+
+void GenerateLabelPrefixRule::HandleSymbol(
+    const verible::Symbol& symbol, const verible::SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (matcher_.Matches(symbol, &manager)) {
```

Comment:
Alternative to consider:

Instead of using a matcher on all begin/end pairs, there are only a few situations that need to be handled, and those can be a switch-case on the node's tag.  For each case, directly access the begin/end labels by descending into the subtree.  There are some functions in CST/statement.h that may help with conditional/loop generates.

That said, as long as the tests pass, I'll accept your choice of implementation.